### PR TITLE
Speed/position filtering, filtering out ITI, speed-related cells

### PR DESCRIPTION
**Pull request by josefbitzenhofer**

### CODE REVIEW

File: viral/models.py [hunk 1]
```diff
@@ -106,3 +107,21 @@ class Cached2pSession(BaseModel):
...
+    rewarded: int
+    trial_frames: np.ndarray
+    iti_start_frame: int
+    iti_end_frame: int
+    frames_positions: np.ndarray
+    frames_speed: np.ndarray
+    corridor_width: int
+    lick_idx: Optional[np.ndarray] = None
+    reward_idx: Optional[np.ndarray] = None
+    signal: np.ndarray
```
Comment (neuromantic99) [inline viral/models.py:124]:
```suggestion
    neural_data: np.ndarray
```

File: viral/models.py [hunk 2]
```diff
@@ -106,3 +107,21 @@ class Cached2pSession(BaseModel):
...
+    trial_end_frame: int
+    rewarded: int
+    trial_frames: np.ndarray
+    iti_start_frame: int
+    iti_end_frame: int
+    frames_positions: np.ndarray
+    frames_speed: np.ndarray
+    corridor_width: int
+    lick_idx: Optional[np.ndarray] = None
+    reward_idx: Optional[np.ndarray] = None
```
Comment (neuromantic99) [inline viral/models.py:123]: I'd probably go with an empty list if there's no licks / rewards rather than an optional type

File: viral/rastermap_utils.py [hunk 3]
```diff
@@ -3,7 +3,9 @@
 import random
 from pathlib import Path
 from scipy.interpolate import interp1d
-from typing import List
+from sklearn.preprocessing import MinMaxScaler
```
Comment (neuromantic99) [inline viral/rastermap_utils.py:6]:
```suggestion
```

File: viral/rastermap_utils.py [hunk 4]
```diff
@@ -15,9 +17,9 @@
...
     get_wheel_circumference_from_rig,
 )
-from viral.models import Cached2pSession
+from viral.models import Cached2pSession, ImagedTrialInfo
 from viral.constants import TIFF_UMBRELLA
-from viral.two_photon import compute_dff, subtract_neuropil, normalize, get_dff
+from viral.two_photon import get_dff
 
 
 def get_spks_pos(s2p_path: Path) -> tuple[np.ndarray]:
```
Comment (neuromantic99) [inline viral/rastermap_utils.py:25]:
```suggestion
def get_spks_pos(s2p_path: Path) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
```

File: viral/rastermap_utils.py [hunk 5]
```diff
@@ -327,45 +360,122 @@ def align_validate_data(spks: np.ndarray, trials: List[TrialInfo]) -> tuple:
...
+
+        # filter out frames with sub-threshold speeds
+        # converting cm/s to cm/frames
+        if filter_speed and speed_threshold is not None:
+            speed_mask = speed[:, 1] >= (speed_threshold)
+            valid_mask |= speed_mask
+        # specified positions that should always be included
+        if filter_position and position_threshold is not None:
+            position_mask = frames_positions[:, 1] >= position_threshold
+            valid_mask |= position_mask  # bitwise OR?
```
[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/rastermap_utils.py:439]: i would start with a vector of ones, then `bitwise and` both masks

File: viral/speed_related_cells.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+def speed_related_cells(spks: np.ndarray, speed: np.ndarray, bin_size: int = 2) -> None:
+    """Plot speed related cells (correlation histograms)"""
+    spks_shuffled = shuffle_neuron_spikes(spks)
+    max_speed = np.max(speed)
+    start = 0
+    end = np.ceil(max_speed / bin_size) * bin_size
+    mean_speed_per_bin = get_mean_value_per_bin(start, end, bin_size)
+    spks_binned = bin_activity(spks, speed, start, end, bin_size)
+    spks_shuffled_binned = bin_activity(spks_shuffled, speed, start, end, bin_size)
```
Comment (neuromantic99) [inline viral/speed_related_cells.py:89]: I wouldn't shuffle the speeds, I would shuffle the binned activity

File: viral/rastermap_utils.py [hunk 7]
```diff
@@ -164,34 +185,36 @@ def get_speed_frame(frame_position: np.ndarray, bin_size: int = 5) -> np.ndarray
             frame_idx (int)
             speed (float)
     """
+    if bin_size <= 1:
+        raise ValueError("You cannot select a bin size <= 1 for speed!")
     n_frames = frame_position.shape[0]
-    n_bins = n_frames // bin_size
+    # Including upper bound
```
[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/rastermap_utils.py:191]: see below

File: viral/rastermap_utils.py [hunk 8]
```diff
@@ -52,7 +64,7 @@ def align_trial_frames(trials: List[TrialInfo], ITI: bool = False) -> np.ndarray
     )
     if not ITI:
         ITI_start_frames = np.array([get_ITI_start_frame(trial) for trial in trials])
-        trial_frames[:, 1] = ITI_start_frames
+        trial_frames[:, 1] = ITI_start_frames - 1
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/rastermap_utils.py:67]: Hmmm not too sure anymore. The ITI_start_frame is the beginning of the ITI hence not really part of the rest of the trial, or am I mistaken?

File: viral/rastermap_utils.py [hunk 9]
```diff
@@ -86,47 +100,44 @@ def get_frame_position(
...
-    ITI: bool = False,
 ) -> np.ndarray:
     """Get the position for every frame and return frames and positions as array.
 
     Args:
         trial (TrialInfo):              A TrialInfo object representing the trial to be analysed.
         trial_frames (np.ndarray):      Numpy array representing the frames in the trial.
         wheel_circumference (float):    The circumference of the wheel on the rig used in centimetres.
-        ITI (bool, optional):           Include the frames in the inter-trial interval (ITI). Defaults to False.
+        ITI (bool, optional):           Whether to include the frames in the inter-trial interval (ITI). Defaults to True.
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/rastermap_utils.py:110]: Sorry, yeah, deleted that bit of the docstring. I got rid of this option as the logic would have to be more complex to deal with the newer sessions which have the position stored in the ITI. And in the code below I have to filter out the ITI from the neural data anyways, hence I did not pursue filtering it out here.

File: viral/rastermap_utils.py [hunk 10]
```diff
@@ -164,34 +185,36 @@ def get_speed_frame(frame_position: np.ndarray, bin_size: int = 5) -> np.ndarray
...
+        else:
+            speed = position_diff / frame_diff
+        if i == len(bins) - 1:
+            speeds[start_idx:] = speed
+        else:
+            speeds[start_idx : bins[i + 1]] = speed
+        # convert from cm / frame to cm / second
+    speeds = speeds * 30
+    speeds[speeds < 0] = 0
+    speeds[speeds > 100] = np.mean(speeds)
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/rastermap_utils.py:216]: Ahm I don't think so. Negative speeds happen here and there, speed going through the roof for some frames happened here and there (although I have "analysed" no more than two sessions.) I built this in after consulting with you as my previous rastermap attempt included MinMaxScaling the speed, hence those outliers were skewing the analysis.

File: viral/rastermap_utils.py [hunk 11]
```diff
@@ -231,64 +254,66 @@ def get_lick_index(trial: TrialInfo) -> np.ndarray | None:
...
-    for start, end in start_end_frames.astype(int):
-        if end > start:
-            segment_frames = np.arange(start, end + 1, 1)
-            reward_segments.append(segment_frames)
-    if len(reward_segments) != 0:
-        rewards = np.concatenate(reward_segments) if reward_segments else None
-        return np.unique(rewards, axis=0)
     else:
-        return None
+        start_frames = np.array(
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/rastermap_utils.py:260]: Not sure if I am not understanding either my code and/or your comment? 
To me it looked like I am defining the "reward_on1" state as start frame and then the "reward_off3" as end frame?
The "allowed_rewards" I built in as you asked for more thorough assertions and tests.

File: viral/rastermap_utils.py [hunk 12]
```diff
@@ -301,9 +326,17 @@ def remap_to_continuous_indices(
...
+    """Return spks/dff, xpos, ypos and trials loaded from cache."""
+
+    class SignalType(Enum):
+        SPKS = "spks"
+        DFF = "dff"
+
+    if signal_type == SignalType.SPKS.value:
+        spks, xpos, ypos = get_spks_pos(s2p_path)
+    elif signal_type == SignalType.DFF.value:
+        spks, xpos, ypos = get_dff_pos(session, s2p_path)
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/rastermap_utils.py:339]: No Enum then 👉🏻👈🏻

File: viral/rastermap_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/rastermap_utils.py:439]: Actually I recall us wanting frames which pass either criterion (either being in locomotion and/or in the "interesting" part of the corridor). But happy to change if you think it would be better?

File: viral/speed_related_cells.py (hunk 6 above)
Comment (josefbitzenhofer) [reply viral/speed_related_cells.py:89]: Maybe it's too late and I should look at it tmrw but where do I shuffle the speeds?
Or do you mean I should shuffle the binned activity rather than shuffling `spks` and then `bin` the `spks_shuffled`?

File: viral/rastermap_utils.py [hunk 13]
```diff
@@ -143,6 +154,16 @@ def get_frame_position(
         fill_value="interpolate",
     )
     interpolated_positions = interp_func(trial_frames)
+
+    if ITI_patch:
+        # If patching is needed, fill up the ITI frames with positions zero
+        ITI_start_frame = get_ITI_start_frame(trial)
+        ITI_end_frame = trial.trial_end_closest_frame
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/rastermap_utils.py:161]: Took me a while but I think I got it now.
Your comment is referring to functions in `cache_2p_sessions.py`.
I discovered a flaw, bear with me:
`process_session` -> `add_imaging_infos_to_trials` -> 1. `add_daq_times_to_trial` 2. `check_timestamps`.
In `add_daq_times_to_trial`, the last frame is determined like this:
```
    # Bit of monkey patching but oh well
    for state in trial.states_info:
        state.start_time_daq = bpod_to_daq(state.start_time).astype(float)
        state.end_time_daq = bpod_to_daq(state.end_time).astype(float)
        state.closest_frame_start = int(
            np.argmin(np.abs(valid_frame_times - state.start_time_daq))
        )
        state.closest_frame_end = int(
            np.argmin(np.abs(valid_frame_times - state.end_time_daq))
        )

        if state.name == "spacer_high_00":
            trial.trial_start_closest_frame = state.closest_frame_start

    last_state = trial.states_info[-1]
    trial.trial_end_closest_frame = last_state.closest_frame_end
```
I.e., the last frame of a trial will be determined as the closest frame to the end_time of the last state of the trial.
In the old version `ITI` is the last state, in the newer version it should be `trigger_panda_ITI`.
Correct me if I am wrong, but I was under the impression that the end of a trial should be the end of the ITI as well?
The flaw was that in `check_timestamps`, the end frame of a trial (and hence the range of frames that were checked) was determined incorrectly like this:
```
    trial_end_state = trial.states_info[-1]

    first_frame_trial = trial_start_state.closest_frame_start
    last_frame_trial = trial_end_state.closest_frame_start
```
I changed it to
```
    first_frame_trial = trial.trial_start_closest_frame
    last_frame_trial = trial.trial_end_closest_frame
```
as we will have the monkey-patched info already.
Hope this resolves the issues, but I think apart from ITI frames not being checked, they should have been included in the output.

File: viral/rastermap_utils.py [hunk 14]
```diff
@@ -164,34 +185,36 @@ def get_speed_frame(frame_position: np.ndarray, bin_size: int = 5) -> np.ndarray
...
-        speeds[start_idx:] = speed
-    result = np.column_stack((frame_position[:, 0], speeds))
-    return result
+    for i in range(len(bins)):
+        start_idx = bins[i]
+        if i == len(bins) - 1:
+            # Handle smaller last bin
+            end_idx = n_frames - 1
+        else:
+            end_idx = bins[i + 1] - 1
```
[the remark this answers is not in the document]
Comment (josefbitzenhofer) [reply viral/rastermap_utils.py:200]: I had to change the logic to being a bit more complex because I was getting out-of-bounds errors for the last bin with your approach

File: viral/rastermap_utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/rastermap_utils.py:67]: yeah i think you're right

File: viral/rastermap_utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (neuromantic99) [reply viral/rastermap_utils.py]: great!
